// @expo/cli/build/src/start/server/middleware/inspector/JsInspector.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    openJsInspector: ()=>openJsInspector,
    closeJsInspector: ()=>closeJsInspector,
    queryInspectorAppAsync: ()=>queryInspectorAppAsync,
    queryAllInspectorAppsAsync: ()=>queryAllInspectorAppsAsync,
    promptInspectorAppAsync: ()=>promptInspectorAppAsync
});
function _chalk() {
    const data = /*#__PURE__*/ _interopRequireDefault(require("chalk"));
    _chalk = function() {
        return data;
    };
    return data;
}
const _launchBrowser = require("./LaunchBrowser");
const _log = require("../../../../log");
const _env = require("../../../../utils/env");
const _prompts = require("../../../../utils/prompts");
const _pageIsSupported = require("../../metro/debugging/pageIsSupported");
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const debug = require("debug")("expo:start:server:middleware:inspector:jsInspector");
let openingBrowserInstance = null;
function openJsInspector(metroBaseUrl, app) {
    if (_env.env.EXPO_USE_UNSTABLE_DEBUGGER) {
        return openExperimentalJsInspector(metroBaseUrl, app);
    } else {
        return openClassicJsInspector(app);
    }
}
async function openExperimentalJsInspector(metroBaseUrl, app) {
    var ref;
    if (!((ref = app.reactNative) == null ? void 0 : ref.logicalDeviceId)) {
        return false;
    }
    const url = new URL("/open-debugger", metroBaseUrl);
    url.searchParams.set("appId", app.description);
    url.searchParams.set("device", app.reactNative.logicalDeviceId);
    url.searchParams.set("target", app.id);
    const response = await fetch(url, {
        method: "POST"
    });
    if (!response.ok) {
        debug("Failed to open React Native DevTools, received response:", response.status);
    }
    return response.ok;
}
/**
 * Chrome DevTools UI implemented for SDK <49.
 * TODO(cedric): Remove this when we fully swap over to the new React Native JS Inspector.
 */ async function openClassicJsInspector(app) {
    _log.Log.log((0, _chalk().default)`{bold Debug:} Opening JavaScript inspector in the browser...`);
    // To update devtoolsFrontendRev, find the full commit hash in the url:
    // https://chromium.googlesource.com/chromium/src.git/+log/refs/tags/{CHROME_VERSION}/chrome/VERSION
    //
    // 1. Replace {CHROME_VERSION} with the target chrome version
    // 2. Click the first log item in the webpage
    // 3. The full commit hash is the desired revision
    const devtoolsFrontendRev = "d9568d04d7dd79269c5a655d7ada69650c5a8336"; // Chrome 100.0.4896.75
    const urlBase = `https://chrome-devtools-frontend.appspot.com/serve_rev/@${devtoolsFrontendRev}/devtools_app.html`;
    const ws = app.webSocketDebuggerUrl.replace(/^ws:\/\//, "");
    const url = `${urlBase}?panel=console&ws=${encodeURIComponent(ws)}`;
    await closeJsInspector();
    openingBrowserInstance = await (0, _launchBrowser.launchInspectorBrowserAsync)(url);
    return true;
}
async function closeJsInspector() {
    await (openingBrowserInstance == null ? void 0 : openingBrowserInstance.close());
    openingBrowserInstance = null;
}
async function queryInspectorAppAsync(metroServerOrigin, appId) {
    const apps = await queryAllInspectorAppsAsync(metroServerOrigin);
    return apps.find((app)=>app.description === appId) ?? null;
}
async function queryAllInspectorAppsAsync(metroServerOrigin) {
    const resp = await fetch(`${metroServerOrigin}/json/list`);
    const apps = transformApps(await resp.json());
    // Only use targets with better reloading support
    return apps.filter((app)=>(0, _pageIsSupported.pageIsSupported)(app));
}
async function promptInspectorAppAsync(apps) {
    var ref;
    if (apps.length === 1) {
        return apps[0];
    }
    // Check if multiple devices are connected with the same device names
    // In this case, append the actual app id (device ID + page number) to the prompt
    const hasDuplicateNames = apps.some((app, index)=>index !== apps.findIndex((other)=>app.deviceName === other.deviceName));
    const choices = apps.map((app)=>{
        const name = app.deviceName ?? "Unknown device";
        return {
            title: hasDuplicateNames ? (0, _chalk().default)`${name}{dim  - ${app.id}}` : name,
            value: app.id,
            app
        };
    });
    const value = await (0, _prompts.selectAsync)((0, _chalk().default)`Debug target {dim (Hermes only)}`, choices);
    return (ref = choices.find((item)=>item.value === value)) == null ? void 0 : ref.app;
}
// The description of `React Native Experimental (Improved Chrome Reloads)` target is `don't use` from metro.
// This function tries to transform the unmeaningful description to appId
function transformApps(apps) {
    const deviceIdToAppId = {};
    for (const app of apps){
        if (app.description !== "don't use") {
            var ref;
            const deviceId = ((ref = app.reactNative) == null ? void 0 : ref.logicalDeviceId) ?? app.id.split("-")[0];
            const appId = app.description;
            deviceIdToAppId[deviceId] = appId;
        }
    }
    return apps.map((app)=>{
        if (app.description === "don't use") {
            var ref;
            const deviceId = ((ref = app.reactNative) == null ? void 0 : ref.logicalDeviceId) ?? app.id.split("-")[0];
            app.description = deviceIdToAppId[deviceId] ?? app.description;
        }
        return app;
    });
}
